screencaster.js

```diff
@@ -0,0 +1,230 @@
+const ws = require("ws");
+const http = require("http");
+const url = require("url");
+const fs = require("fs");
+
+const SingleBrowserImplementation = require("puppeteer-cluster/dist/concurrency/SingleBrowserImplementation").default;
+
+const indexHTML = fs.readFileSync("/app/screencast/index.html", {encoding: "utf8"});
+
+
+// ===========================================================================
+class ScreenCaster
+{
+  constructor(cluster, port) {
+    this.cluster = cluster;
+
+    this.httpServer = http.createServer((req, res) => {
+      const pathname = url.parse(req.url).pathname;
+      if (pathname === "/") {
+        res.writeHead(200, {"Content-Type": "text/html"});
+        res.end(indexHTML);
+      } else {
+        res.writeHead(404, {"Content-Type": "text/html"});
+        res.end("Not Found");
+      }
+    });
+
+    this.allWS = new Set();
+
+    this.targets = new Map();
+    this.caches = new Map();
+    this.urls = new Map();
+
+    this.wss = new ws.Server({ noServer: true });
+
+    this.wss.on("connection", (ws) => this.initWS(ws));
+
+    this.httpServer.on("upgrade", (request, socket, head) => {
+      const pathname = url.parse(request.url).pathname;
+
+      if (pathname === "/ws") {
+        this.wss.handleUpgrade(request, socket, head, (ws) => {
+          this.wss.emit("connection", ws, request);
+        });
+      }
+    });
+
+    this.httpServer.listen(port);
+    console.log(`Screencast Server started on: ${port}`);
+  }
+
+  initWS(ws) {
+    for (const id of this.targets.keys()) {
+      const data = this.caches.get(id);
+      const url = this.urls.get(id);
+      const msg = {"msg": "newTarget", id, url, data};
+      ws.send(JSON.stringify(msg));
+    }
+
+    this.allWS.add(ws);
+
+    if (this.allWS.size === 1) {
+      this.startCastAll();
+    }
+
+    ws.on("close", () => {
+      console.log("ws closed");
```

Comment:
I would expand the 'ws' to be a big clearer. 'web server' ect. I think that would be helpful